Add ELMo-style scalar mixing to BERT wrapper

iftenney opened the pull request:
Per suggestion of BERT authors, this makes everything dramatically better when using BERT-as-a-feature. Results coming soon.

pep8speaks commented:
Hello @iftenney! Thanks for submitting the PR.

 - In the file [`src/bert/utils.py`](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py), following are the PEP8 issues :

> [Line 13:1](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L13): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 39:5](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L39): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 40:38](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L40): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 40:39](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L40): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 52:80](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 85:80](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 100:40](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L100): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 111:1](https://github.com/jsalt18-sentence-repl/jiant/blob/a80c6ced7592e3d7f862c2d8b742081e6b7dfd4b/src/bert/utils.py#L111): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

sleepinyourhat commented on src/bert/utils.py:
```diff
@@ -24,6 +26,16 @@ def __init__(self, args, cache_dir=None):
         for param in self.model.parameters():
             param.requires_grad = bool(args.bert_fine_tune)
 
+        # Configure scalar mixing, ELMo-style
```
Make this a warning?

iftenney replied:
Made an assert.

sleepinyourhat commented on config/defaults.conf:
```diff
@@ -216,6 +216,8 @@ bert_embeddings_mode = "none"  // How to handle the embedding layer of the
                                //   lexical layer
                                // "only" returns only lexical layer
+                               // "mix" uses ELMo-style scalar mixing (with
+                               // learned weights) across all layers
```
Make it clear what one should do to run BERT-style fine-tuning?

iftenney replied:
That's handled by the `bert_fine_tune` parameter below.

sleepinyourhat replied:
Er, sorry: What should one set here to follow the standard BERT protocol (when bert_fine_tune = 1). None? Does it not matter?

sleepinyourhat replied:
Clarify and merge?

sleepinyourhat commented on src/bert/utils.py:
```diff
@@ -24,6 +26,16 @@ def __init__(self, args, cache_dir=None):
+        if self.embeddings_mode == "mix":
+            num_layers = self.model.config.num_hidden_layers
+            self.scalar_mix = scalar_mix.ScalarMix(num_layers + 1,
+                                                   do_layer_norm=False)
```
Are gradients to these handled in the same way as ELMo scalar gradients? (i.e., trained even when the rest of BERT isn't?)

iftenney replied:
Yes, that's the intent. Added comment to clarify.

sleepinyourhat replied:
Cool. I'll assume you have a way to double check this.

sleepinyourhat commented:
Still not my intended question. Rephrasing: If you want to use jiant to
exactly reproduce one of the experiments in the BERT paper, how should you
set bert_embeddings_mode?

On Wed, Jan 30, 2019 at 11:25 PM Ian Tenney <notifications@github.com>
wrote:

> *@iftenney* commented on this pull request.
> ------------------------------
>
> In config/defaults.conf
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_jsalt18-2Dsentence-2Drepl_jiant_pull_480-23discussion-5Fr252531325&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=sCzLyHdE8zgQwk2-sKwA1w&m=qWo0YcVu2GNwL2QFONiaViKPjBrOJF4FiiEYWmOqWaE&s=8A1FlG_iP8LWwqYTThSpsVOJGxL06Qy-wfEeRNpqhl8&e=>
> :
>
> > @@ -216,6 +216,8 @@ bert_embeddings_mode = "none"  // How to handle the embedding layer of the
>                                 // "cat" returns top-layer concatenated with
>                                 //   lexical layer
>                                 // "only" returns only lexical layer
> +                               // "mix" uses ELMo-style scalar mixing (with
> +                               // learned weights) across all layers
>
> They're orthogonal - setting bert_fine_tune = 1 makes all the BERT
> weights trainable, while bert_embeddings_mode just controls which
> activations are returned by the encoder module.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_jsalt18-2Dsentence-2Drepl_jiant_pull_480-23discussion-5Fr252531325&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=sCzLyHdE8zgQwk2-sKwA1w&m=qWo0YcVu2GNwL2QFONiaViKPjBrOJF4FiiEYWmOqWaE&s=8A1FlG_iP8LWwqYTThSpsVOJGxL06Qy-wfEeRNpqhl8&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABOZWZwXUc1INQisvbSCMEmAJYKbttnZks5vInAxgaJpZM4abcEF&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=sCzLyHdE8zgQwk2-sKwA1w&m=qWo0YcVu2GNwL2QFONiaViKPjBrOJF4FiiEYWmOqWaE&s=phe1ZB-GXKRTHvXFzTFMWL46Vx88mQM_gc6eKwgFQ3Y&e=>
> .
>

iftenney commented:
@sleepinyourhat : Sorry, didn't understand what you were asking before. `bert_embeddings_mode=none` should return just the top layer, which is the standard behavior.

Updated comments and added "top" as a synonym.

sleepinyourhat commented:
Tx!
